packages/astro/src/assets/build/generate.ts

```diff
@@ -53,113 +72,176 @@ export async function generateImage(
 		clientRoot = config.outDir;
 	}
 
-	const isLocalImage = isESMImportedImage(options.src);
+	return {
+		logger,
+		count,
+		useCache,
+		assetsCacheDir,
+		serverRoot,
+		clientRoot,
+		imageConfig: config.image,
+		assetsFolder: config.build.assets,
+	};
+}
 
-	const finalFileURL = new URL('.' + filepath, clientRoot);
-	const finalFolderURL = new URL('./', finalFileURL);
+export async function generateImagesForPath(
+	originalFilePath: string,
+	transforms: MapValue<AssetsGlobalStaticImagesList>,
+	env: AssetEnv,
+	queue: PQueue
+) {
+	const originalImageData = await loadImage(originalFilePath, env);
```

Comment:
In theory, the fact that we load every image at the beginning even when all the transforms might be cached already is slower, especially for fully cached builds. However, in practice I found that this didn't change much, probably that the rest of the improvements counteract this, or maybe hard drives are just fast enough now.

If we want to get it of this, what we could probably do is store a list of cached hashes and check first if all the transforms requested are already cached before loading the image, shouldn't be too hard.

Reply:
I think we should apply your suggestions, probably in a follow-up PR. An I/O operation is usually more expensive than a memory operation, and even though Node.js is really great in I/O operations, I think we should try to avoid them if we can.

Reply:
I definitely agree, but I think it's not as bad as it might seems because the file is already written in the same process, so it's presumably already in some sort of hot cache.

Reply:
Maybe we also make `originalImageData` a function to lazily load it, and those who call in parallel would share the same cache? Vite has this [utility](https://github.com/vitejs/vite/blob/9eec4fb3515c797d18686575c3e644da224657f0/packages/vite/src/node/plugins/css.ts#L1232-L1243) (which could be generalized) to simplify a bit.

But also I think we can do it in a follow-up too if there's not a big perf impact currently.